File: day1/go/day1.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

// ReadInts reads whitespace-separated ints from r. If there's an error, it
// returns the ints successfully read so far as well as the error value.
// Borrowed from https://stackoverflow.com/questions/9862443/golang-is-there-a-better-way-read-a-file-of-integers-into-an-array
func ReadInts(r io.Reader) ([]int, error) {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanWords)
	var result []int
	for scanner.Scan() {
		x, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return result, err
		}
		result = append(result, x)
	}
	return result, scanner.Err()
}

// Part 1
func CountIncreases(numbers []int) {
	num_increases := 0
	for idx, _ := range numbers {
		if idx == 0 {
			continue
		}
		if numbers[idx] > numbers[idx-1] {
			num_increases++
		}
	}
	fmt.Println("Num increases: ", num_increases)
}

// Part 2
func Sum(array []int) int {
	result := 0
	for i := 0; i < len(array); i++ {
		result += array[i]
	}
	return result
}

func SlidingWindow(numbers []int, window_size int) {

	window_sums := []int{}

	for idx, _ := range numbers {
		if idx < window_size-1 {
			continue
		}

		slice := numbers[(idx - window_size + 1):(idx + 1)]

		window_sums = append(window_sums, Sum(slice))

	}
	CountIncreases(window_sums)
}

func main() {
	// Read in data
	file, _ := os.Open("day1/inputs/input1.txt")
	ints, _ := ReadInts(file)

	// Q1
	CountIncreases(ints)

	// Q2
	SlidingWindow(ints, 3)

}

File: day1/python/day1.py
def part1(puzzle_input):
    """Count the number of increases in input"""
    n_increases = 0
    
    for prev, next in zip(puzzle_input[:-1], puzzle_input[1:]):
        if next > prev:
            n_increases +=1 
    return n_increases

def part2(puzzle_input, window_size):
    """Sliding window of n: 
    - Create sum of n vector
    - Feed into part1
    """
    window_sums = list()
    for i in range(len(puzzle_input)-window_size + 1):
        window = puzzle_input[i : i + window_size]
        window_sums.append(sum(window))
    return part1(window_sums)
        

if __name__=="__main__":
    with open("day1/inputs/input1.txt", "r") as f:
        puzzle_input = f.readlines()
        puzzle_input = [int(t) for t in puzzle_input]

    with open("day1/inputs/test1.txt", "r") as f:
        test1 = f.readlines()
        test1 = [int(t) for t in test1]

    part1_answer = part1(puzzle_input)
    print(f"Part 1: {part1_answer}")

    part2_answer = part2(puzzle_input, window_size=3)
    print(f"Part 2: {part2_answer}")